Preparing change log for beta release (#147)
silverstripe/developer-docs: en/04_Changelogs/5.0.0.md

# 5.0.0 (unreleased)

> Silverstripe CMS 5.0.0 is the first major release since November 2017. It's also the first major release since we adopted a formal [Major Release Policy](/Project_governance/Major_release_policy).

Author: ```suggestion
Silverstripe CMS 5.0.0 is the first major release since November 2017. It's also the first major release since the adoption of a formal [Major Release Policy](/Project_governance/Major_release_policy).
```

Reviewer: You didn't make this change... was that intentional?

Reviewer: No response so I'm marking unresolved.

> Silverstripe CMS 5.0.0 is the first major release since November 2017. It's also the first major release since we adopted a formal [Major Release Policy](/Project_governance/Major_release_policy).

Reviewer: ```suggestion
Silverstripe CMS 5.0.0 is the first major release since November 2017. It's also the first major release since we adopted a formal [Major Release Policy](/project_governance/major_release_policy/).
```
URLs are normalised to lowercase in the docs site.

Reviewer: I won't explicitly call out the other links that need updating below because it could get spammy and I'd probably miss some anyway - but this applies to all relative links.

Author: Done. I went through all of the links and I think I got them all lowercased.

This release contains many breaking changes and dependency upgrades. We tried our best to minimise upgrades woes, but you should allocate additional time to account for regression testing.

> This changelogs provides an exhaustive list changes between Silverstripe CMS 4 and 5.

Author: ```suggestion
This changelogs provides an exhaustive list of changes between Silverstripe CMS 4 and 5.
```

> This changelogs provides an exhaustive list changes between Silverstripe CMS 4 and 5.

Reviewer: ```suggestion
This changelog provides an exhaustive list changes between Silverstripe CMS 4 and 5.
```
I'm on the fence about this comment - we do our best, but someone will undoubtedly end up finding _something_ we haven't explicitly called out. Unless you include the commits list at the bottom as what makes it exhaustive?

Author: Taking the "exhaustive" out.

Author: Done

## Overview

- [Change to commercially supported modules](#supported-modules)

> - [Fixed dependencies](#fixed-depdencies)

Author: ```suggestion
- [Change to commercially supported modules](#supported-modules)
  - [Module not supported in Silverstripe CMS 5](#unsupported-modules)
- [Fixed dependencies](#fixed-depdencies)
```

Reviewer: You suggested this change and then resolved it..... does that mean you now don't think it should be there? Or did you think you added it?

I don't mind either way, whether it's there or not, just wanted to make sure you had done the thing you intended to do.

Author: I think I most over-ridden one of my own commit at some stage or something, because there's a bunch of things I committed directly from GitHub that aren't there anymore.

> - [Fixed dependencies](#fixed-depdencies)

Reviewer: ```suggestion
- [Fixed dependencies](#fixed-dependencies)
```

# 5.0.0 (unreleased)

Silverstripe CMS 5.0.0 is the first major release since November 2017. It's also the first major release since the adoption of a formal [Major Release Policy](/Project_governance/Major_release_policy).

This release contains many breaking changes and dependency upgrades. We tried our best to minimise upgrades woes, but you should allocate additional time to account for regression testing.

> This changelogs provides an exhaustive list of changes between Silverstripe CMS 4 and 5.

Reviewer 2: ```suggestion
This changelog provides an exhaustive list of changes between Silverstripe CMS 4 and 5.
```

## Overview

- [Change to commercially supported modules](#supported-modules)

>   - [Module not supported in Silverstripe CMS 5](#unsupported-modules)

Reviewer 2: ```suggestion
  - [Modules not supported in Silverstripe CMS 5](#unsupported-modules)
```

  - [Other features](#other-features)
- [Bugfixes](#bugfixes)

## Change to supported modules {#supported-modules}

Some Silverstripe CMS modules are commercially supported. Silverstripe commits to looking after those modules for the duration of the Silverstripe CMS 5 lifecycle.

Review the list of [Silverstripe CMS 5 supported modules](/Project_Governance/Supported_Modules).

### Module not supported in Silverstripe CMS 5

> Some modules that were supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5. Just because a module is not supported officially, doesn't mean that you shouldn't be using it. However, it does mean that it's more likely to suffer from regressions. You should take this in consideration when choosing to install an unsupported module in your project.

Reviewer 2: I think saying that it's more likely modules will suffer from regressions might be scaring people unnecessarily. Can we find better wording? Because if no one touches the code then how would a regression occur?

Author: Even if no one touches the code, it is still possible that later change in a CMS 5 minor will break things.

I'm happy to rethink this wording, but we shouldn't say that running unsupported module is just a safe as running supported ones.

Reviewer: I think it does the community an injustice to effectively say "all community modules are less safe than the ones _we_ look after"

Some of those modules will provide CMS-5-compatible versions. Others will be dropped altogether.

> Email community@silverstripe.org ff you are keen to maintain some of the module that will no longer be officially supported.

Reviewer 2: ```suggestion
Email community@silverstripe.org if you are keen to maintain some of the modules that will no longer be officially supported.
```

[composer/installers](https://github.com/composer/installers) | Dropped | Composer is more established than Silverstripe CMS.
[cwp-themes/default](https://github.com/silverstripe/cwp-theme-default) | Dropped | 
[cwp/cwp](https://packagist.org/packages/cwp/cwp) | CMS5 compatible without support | 
[cwp/cwp-core](https://packagist.org/packages/cwp/cwp-core) | CMS5 comptabile without support | 
[cwp/cwp-pdfexport](https://packagist.org/packages/cwp/cwp-pdfexport) | Dropped | 
[cwp/cwp-search](https://packagist.org/packages/cwp/cwp-search) | CMS5 comptabile without support | 
[hafriedlander/phockito](https://github.com/hafriedlander/phockito) | Dropped | 
[hafriedlander/silverstripe-phockito](https://packagist.org/packages/hafriedlander/silverstripe-phockito) | Dropped | 
[lekoala/silverstripe-debugbar](https://packagist.org/packages/lekoala/silverstripe-debugbar) | CMS5 comptabile without support | Debug bar is a development tool that should not be installed in production.
[silverstripe/activedirectory](https://packagist.org/packages/silverstripe/activedirectory) | Dropped | 

> [silverstripe/akismet](https://packagist.org/packages/silverstripe/akismet) | Dropped | The parent library is outdated and they are better alternatives like [UndefinedOffset/silverstripe-nocaptcha](https://packagist.org/packages/undefinedoffset/silverstripe-nocaptcha)

Reviewer 2: ```suggestion
[silverstripe/akismet](https://packagist.org/packages/silverstripe/akismet) | Dropped | The parent library is outdated and there are better alternatives like [UndefinedOffset/silverstripe-nocaptcha](https://packagist.org/packages/undefinedoffset/silverstripe-nocaptcha)
```

[silverstripe/behat-extension](https://github.com/silverstripe/silverstripe-behat-extension) | CMS5 comptabile without official support | 

> [silverstripe/ckan-registry](https://packagist.org/packages/silverstripe/ckan-registry) | CMS5 comptabile without support | 

Reviewer 2: Is there a difference between 'without official support' and 'without support'?

Author: No there isn't. That's typo.

[silverstripe/comment-notifications](https://packagist.org/packages/silverstripe/comment-notifications) | CMS5 comptabile without official support | 
[silverstripe/comments](https://packagist.org/packages/silverstripe/comments) | CMS5 comptabile without official support | 
[silverstripe/content-widget](https://packagist.org/packages/silverstripe/content-widget) | CMS5 comptabile without support | 
[silverstripe/fulltextsearch](https://packagist.org/packages/silverstripe/fulltextsearch) | CMS5 comptabile without support | 
[silverstripe/graphql-devtools](https://packagist.org/packages/silverstripe/graphql-devtools) | CMS5 comptabile without official support | 

> [silverstripe/html5](https://packagist.org/packages/silverstripe/html5) | Folded back in core | 

Reviewer 2: Should we be explicit whether core means Framework or CMS?

Author: That's a good idea. I'll do it.

Author: None of those modules were core modules. I've created a dedicated "supported module" page and link to it. The list of modules that has been split between core and optional modules.

[silverstripe/postgresql](https://packagist.org/packages/silverstripe/postgresql) | CMS5 comptabile without support | 
[silverstripe/recipe-ccl](https://github.com/silverstripe/recipe-ccl) | CMS5 comptabile without support | 
[silverstripe/recipe-solr-search](https://github.com/silverstripe/recipe-solr-search) | CMS5 comptabile without support | 
[silverstripe/spellcheck](https://packagist.org/packages/silverstripe/spellcheck) | Dropped | 
[silverstripe/sqlite3](https://packagist.org/packages/silverstripe/sqlite3) | CMS5 comptabile without support | 
[silverstripe/widgets](https://packagist.org/packages/silverstripe/widgets) | CMS5 comptabile without official support | 
[tijsverkoyen/akismet](https://github.com/tijsverkoyen/akismet) | Dropped | 
[tractorcow/classproxy](https://github.com/tractorcow/classproxy) | CMS5 comptabile without official support | 
[tractorcow/silverstripe-proxy-db](https://packagist.org/packages/tractorcow/silverstripe-proxy-db) | CMS5 comptabile without official support | 

> ## Fixed dependencies {#fixed-depdencies}

Reviewer 2: Remove until needed so we don't forget it here empty?

```suggestion
```

Author: I'll add a link to the full list below.

Email [community@silverstripe.org](mailto:community@silverstripe.org) if you are keen to help maintain some of the modules that will no longer be officially supported.

Module | Status | Notes
-- | -- | --
[bringyourownideas/silverstripe-composer-security-checker](https://packagist.org/packages/bringyourownideas/silverstripe-composer-security-checker) | Dropped | Underlying service has been discontinued.
[composer/installers](https://packagist.org/packages/composer/installers) | Dropped | Composer is more established than Silverstripe CMS.
[cwp-themes/default](https://packagist.org/packages/cwp-themes/default) | Dropped | 
[cwp/cwp](https://packagist.org/packages/cwp/cwp) | CMS5 compatible without commercial support | 
[cwp/cwp-core](https://packagist.org/packages/cwp/cwp-core) | CMS5 compatible without commercial support | 
[cwp/cwp-pdfexport](https://packagist.org/packages/cwp/cwp-pdfexport) | Dropped | 
[cwp/cwp-search](https://packagist.org/packages/cwp/cwp-search) | CMS5 compatible without commercial support | 

> [hafriedlander/phockito](https://github.com/hafriedlander/phockito) | Dropped | 

Reviewer: ```suggestion
[hafriedlander/phockito](https://packagist.org/packages/hafriedlander/phockito) | Dropped | 
```

[silverstripe/behat-extension](https://packagist.org/packages/silverstripe/behat-extension) | CMS5 compatible without official support | 
[silverstripe/ckan-registry](https://packagist.org/packages/silverstripe/ckan-registry) | CMS5 compatible without commercial support | 
[silverstripe/comment-notifications](https://packagist.org/packages/silverstripe/comment-notifications) | CMS5 compatible without official support | 
[silverstripe/comments](https://packagist.org/packages/silverstripe/comments) | CMS5 compatible without official support | 
[silverstripe/content-widget](https://packagist.org/packages/silverstripe/content-widget) | CMS5 compatible without commercial support | 
[silverstripe/fulltextsearch](https://packagist.org/packages/silverstripe/fulltextsearch) | CMS5 compatible without commercial support | 
[silverstripe/graphql-devtools](https://packagist.org/packages/silverstripe/graphql-devtools) | CMS5 compatible without official support | 
[silverstripe/html5](https://packagist.org/packages/silverstripe/html5) | Folded back in core | 
[silverstripe/postgresql](https://packagist.org/packages/silverstripe/postgresql) | CMS5 compatible without commercial support | 

> [silverstripe/recipe-ccl](https://github.com/silverstripe/recipe-ccl) | CMS5 compatible without commercial support | 

Reviewer: ```suggestion
[silverstripe/recipe-ccl](https://packagist.org/packages/silverstripe/recipe-ccl) | CMS5 compatible without commercial support | 
```

  - [URLs generated by Silverstripe CMS don't include a trailing slash by default](#trailing-slash)
  - [CWP Agency Extensions](#cwp-agency-extensions)
- [Features and enhancements](#features-and-enhancements)
  - [Extension changes](#extension-changes)
  - [Other new features](#other-features)
- [Bugfixes](#bugfixes)

## Change to supported modules {#supported-modules}

Some Silverstripe CMS modules are commercially supported. Silverstripe commits to looking after those modules for the duration of the Silverstripe CMS 5 lifecycle.

Review the list of [Commercially Supported Modules for Silverstripe CMS 5](/project_governance/supported_modules).

> ### Community supported modules in Silverstripe CMS 5

Reviewer: https://github.com/silverstripe/developer-docs/pull/147#discussion_r1099560123
Hmmmm I don't think this rename matches what the section is about. It's really about modules which used to be commercially supported. This heading makes me think it's going to talk about my random module which silverstripe doesn't know exists.

Please change this back - or otherwise keep it clearly on subject (e.g. "Previously commercially supported modules" or "Dropping commercial support for some modules").

I think the changes you've made in the _body_ of the text satisfy the discussion in https://github.com/silverstripe/developer-docs/pull/147#discussion_r1095165654

Some modules that were commercially supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5. Some of those modules will provide CMS5-compatible versions. Others will be dropped altogether.

Just because a module is not "commercially supported", doesn't mean that you shouldn't be using it. Community supported modules are maintained on a "best-effort" basis. You should take this into consideration when choosing to install a community supported module in your project.

Email [community@silverstripe.org](mailto:community@silverstripe.org) if you are keen to help maintain some of the modules that will no longer be commercially supported.

> Module | Status | Notes

Reviewer: Please add [dnadesign/silverstripe-elemental-subsites](https://packagist.org/packages/dnadesign/silverstripe-elemental-subsites) to this list.

[undefinedoffset/sortablegridfield](https://packagist.org/packages/undefinedoffset/sortablegridfield) | Dropped | [symbiote/silverstripe-gridfieldextensions](https://packagist.org/packages/symbiote/silverstripe-gridfieldextensions) provides equivalent functionality.

## Fixed dependencies

Silverstripe CMS relies on third party dependencies to implement many features. We have defined several *fixed dependencies* to provide certainty to Silverstripe CMS project owners and minimise the risk of composer conflicts. These dependencies will not be upgraded to new major versions within the Silverstripe CMS 5 release line.

Review [Silverstripe CMS 5 fixed dependencies](/project_governance/fixed_dependencies) for more details.

## API changes {#api-changes}

This is a major release and contains many breaking API changes. Deprecation warnings have been added to the latest Silverstripe CMS 4 release to advise you of APIs that have been removed in Silverstripe CMS 5.

[Enable deprecation warnings](/upgrading/deprecations) prior to upgrading to Silverstripe CMS 5 to see if your project relies on any APIs that are no longer unavailable.

> ### Removed API (by module, alphabetically) {#api-removed}

Author: Didn't add any API links under this header because there won't be anything to link to since the APIs are gone.

Reviewer: That's good, because links under this heading will be automatically generated in a separate PR.

  - [Other features](#other-features)
- [Bugfixes](#bugfixes)

## Change to supported modules {#supported-modules}

Some Silverstripe CMS modules are commercially supported. Silverstripe commits to looking after those modules for the duration of the Silverstripe CMS 5 lifecycle.

Review the [list Commercially Supported Modules for Silverstripe CMS 5](/Project_Governance/Supported_Modules).

> ### Module not supported in Silverstripe CMS 5

Author: ```suggestion
### Module not supported in Silverstripe CMS 5 {#unsupported-modules}
```

Reviewer: You didn't make this change - was that intentional?

Reviewer: No response on this so I've marked it unresolved.

Author: Header got renamed to "Community supported in Silverstripe CMS 5"

Reviewer: Please add the word "modules" in there

Reviewer 3: Hmmmm I don't think this rename matches what the section is about. It's really about modules which _used_ to be commercially supported. This heading makes me think it's going to talk about my random module which silverstripe doesn't know exists.

Please change this back - or otherwise keep it clearly on subject (e.g. "Previously commercially supported modules" or "Dropping commercial support for some modules")

Some modules that were supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5. 

> Just because a module is not supported officially, doesn't mean that you shouldn't be using it. However, it does mean that it's more likely to suffer from regressions. You should take this in consideration when choosing to install an unsupported module in your project.

Author: ```suggestion
Just because a module is not supported officially, doesn't mean that you shouldn't be using it. Unsupported modules are maintained on a "best-effort" basis. You should take this into consideration when choosing to install an unsupported module in your project.
```

Reviewer: You didn't make this change. Was that intentional? I think this is an important change, so I'm making it as unresolved.

Author: I think I change something that "outdated" this comment. I'll bring it back.

[silverstripe/widgets](https://packagist.org/packages/silverstripe/widgets) | CMS5 compatible without official support | 
[tijsverkoyen/akismet](https://github.com/tijsverkoyen/akismet) | Dropped | 
[tractorcow/classproxy](https://github.com/tractorcow/classproxy) | CMS5 compatible without official support | 
[tractorcow/silverstripe-proxy-db](https://packagist.org/packages/tractorcow/silverstripe-proxy-db) | CMS5 compatible without official support |
[undefinedoffset/sortablegridfield](https://packagist.org/packages/undefinedoffset/sortablegridfield) | Dropped | [symbiote/silverstripe-gridfieldextensions](https://packagist.org/packages/symbiote/silverstripe-gridfieldextensions) provides equivalent functionality.

## Fixed dependencies {#fixed-depdencies}

Silverstripe CMS relies on many third party dependencies to implement various features. Silverstripe CMS defines many *fixed dependencies* that provide certainty to Silverstripe CMS projects and minimise the risk of composer conflicts.

We defined a list of *fixed dependencies* for the lifetime of Silverstripe CMS 5. Those dependencies will not be upgraded to new major version within the Silverstripe CMS 5 release line.

> Review Silverstripe CMS 5 fixed dependencies for more details.

Author: ```suggestion
Review [Silverstripe CMS 5 fixed dependencies](/Project_Governance/Fixed_dependencies) for more details.
```

Reviewer: Marking unresolved as it wasn't done and if you're keeping this sentence it should have the link.

Author: I'll add this.

## API changes {#api-changes}

> This is a major release and as a result there are a number of breaking API changes. For a full list of these see [upgrading your project](/upgrading/upgrading_your_project). Some specific details about a few of them are below.

Author: My view is that the full list should be here.

Author: I moved some of that language to the top of the changelog.

Reviewer: I disagree - I think we should move the full list of changes into a separate document.
If they're all in the main changelog, people have to scroll past a GIANT WALL of text to see the relevant changes. It's going to be a much bigger wall once Steve has finished auto-generating the full list of changes. I think we should move all of the "removed this deprecated class, changed this to private" etc into a separate document that people can look at _if they're interested and find that relevant_.

The large majority of people will either not look at it at all, or will only search for specific things that are relevant to their project. Having it in the main changelog makes people less likely to bother scrolling to see the changes we have chosen to explicitly call out in our hand-written bits.

The actual URL will depend on where Steve ends up putting the list.

Reviewer: That said, we can probably leave that for Steve's PR - so I'm okay with removing it from here for now.

Author: The approach we've generally taken over the years is to put everything in one file. Yes, it's a big wall of text. My instinct would be to ship it like that for the beta and ask devs if they have a preference. We can re-adjust for stable.

Reviewer: We've not done this before - not since CMS 4.0.0. I think it's okay for us to change an approach for something like this.

> ask devs if they have a preference

Do we have a clear list of who we're asking and how we're asking them - and when? And when we'll action the response?

Author: I don't think we have to make this a massive consultation. Let's maybe round up a dozen devs and see what they think. We'll be doing that anyway in our bespoke/community outreach.

Reviewer: Okay - so long as there _is_ a consultation, if that's what we're deciding on. 👍

Author: https://github.com/silverstripe/developer-docs/issues/151

> This is a major release and contains many breaking API changes.

Author: ```suggestion
This is a major release and contains many breaking API changes. Deprecation warnings have been added to the latest Silverstripe CMS 4 release to advise you of APIs that have been removed in Silverstripe CMS 5.

[Enable deprecation warnings](/Upgrading/Deprecations) prior to upgrading to Silverstripe CMS 5 to see if your project relies on any APIs that are no longer unavailable.
```

  - [Other features](#other-features)
- [Bugfixes](#bugfixes)

## Change to supported modules {#supported-modules}

Some Silverstripe CMS modules are commercially supported. Silverstripe commits to looking after those modules for the duration of the Silverstripe CMS 5 lifecycle.

> Review the [list Commercially Supported Modules for Silverstripe CMS 5](/Project_Governance/Supported_Modules).

Reviewer: ```suggestion
Review the [list of Commercially Supported Modules for Silverstripe CMS 5](/Project_Governance/Supported_Modules).
```

> ### Module not supported in Silverstripe CMS 5

Reviewer: ```suggestion
### Modules not supported in Silverstripe CMS 5
```

> Some modules that were supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5. 

Reviewer: ```suggestion
Some modules that were commercially supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5. 
```

Just because a module is not supported officially, doesn't mean that you shouldn't be using it. However, it does mean that it's more likely to suffer from regressions. You should take this in consideration when choosing to install an unsupported module in your project.

> Some of those modules will provide CMS-5-compatible versions. Others will be dropped altogether.

Reviewer: ```suggestion
Some of those modules will provide CMS5-compatible versions. Others will be dropped altogether.
```
or `CMS 5 compatible`

> Some of those modules will provide CMS-5-compatible versions. Others will be dropped altogether.

Reviewer: I think this should be added to the end of the "Some modules that were commercially supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5." paragraph above.

Author: Will do

> Email [community@silverstripe.org](mailto:community@silverstripe.org) if you are keen to help maintain some of the modules that will no longer be officially supported.

Reviewer: Who will be responsible for vetting these and providing access, etc?

Author: That's already set up and I'm getting the messages.

Reviewer: I'm going to take that to mean that you will be the one responsible for that?

Author: Yes.

Module | Status | Notes
-- | -- | --

> [bringyourownideas/silverstripe-composer-security-checker](https://packagist.org/packages/bringyourownideas/silverstripe-composer-security-checker) | Dropped | Underlying service has been discontinued.

Reviewer: You have a mix of packagist and github URLs here - better to just pick one or the other.

Author: I'll update everything to point to packagist

Author: Done

> [composer/installers](https://github.com/composer/installers) | Dropped | Composer is more established than Silverstripe CMS.

Reviewer: It's weird mentioning `composer/installers` here - both because we never were _really_ supporting it anyway, and because saying "dropped" about this makes it sound like somehow its value is diminished (that's certainly the case for all of the others here that were dropped).

I'm on the fence between just not mentioning it, or giving it some unique status, or mentioning is outside of the table like "This was on our list of commercially supported dependencies, but in reality because Composer is more established than Silverstripe CMS we have never needed to provide any support for it. We have decided to remove it from the list of supported dependencies for CMS 5 to reduce confusion on this point."

Author: It was on the list of supported module ... for some ungodly reason I never understood.

It was on the list, so we are technically removing something.

That's why I added the "Notes" column at the end.

Reviewer: Right. But as I mentioned, "dropped" has a finality to it that just shouldn't be there for this particular scenario. If you want to keep it in the list with "dropped" I think the notes need to be a lot clearer (the text I provided for "outside the table" would also work as a note for this one I guess)

Reviewer: Can you please update the note to be clearer as requested, or give a reason why you aren't doing that?

Author: Added a more detailed explanation and change the "dropped" on that line to "removed".

> [cwp-themes/default](https://github.com/silverstripe/cwp-theme-default) | Dropped | 

Reviewer: As an aside, we should remove all of the "dropped" ones from kitchen sink immediately, and remove any "compatible but without support" from the sink after or around the RC (so we can know they're compatible but thereafter not fail the whole sink for them)

Author: Agree. https://github.com/silverstripe/recipe-kitchen-sink/issues/29

[cwp/cwp](https://packagist.org/packages/cwp/cwp) | CMS5 compatible without support | 
[cwp/cwp-core](https://packagist.org/packages/cwp/cwp-core) | CMS5 compatible without support | 
[cwp/cwp-pdfexport](https://packagist.org/packages/cwp/cwp-pdfexport) | Dropped | 
[cwp/cwp-search](https://packagist.org/packages/cwp/cwp-search) | CMS5 compatible without support | 
[hafriedlander/phockito](https://github.com/hafriedlander/phockito) | Dropped | 
[hafriedlander/silverstripe-phockito](https://packagist.org/packages/hafriedlander/silverstripe-phockito) | Dropped | 
[lekoala/silverstripe-debugbar](https://packagist.org/packages/lekoala/silverstripe-debugbar) | CMS5 compatible without support | Debug bar is a development tool that should not be installed in production.
[silverstripe/activedirectory](https://packagist.org/packages/silverstripe/activedirectory) | Dropped | 

> [silverstripe/akismet](https://packagist.org/packages/silverstripe/akismet) | Dropped | The parent library is outdated and they are better alternatives like [UndefinedOffset/silverstripe-nocaptcha](https://packagist.org/packages/undefinedoffset/silverstripe-nocaptcha)

Reviewer: ```suggestion
[silverstripe/akismet](https://packagist.org/packages/silverstripe/akismet) | Dropped | The parent library is outdated and there are better alternatives like [UndefinedOffset/silverstripe-nocaptcha](https://packagist.org/packages/undefinedoffset/silverstripe-nocaptcha)
```

[silverstripe/behat-extension](https://github.com/silverstripe/silverstripe-behat-extension) | CMS5 compatible without official support | 

> [silverstripe/ckan-registry](https://packagist.org/packages/silverstripe/ckan-registry) | CMS5 compatible without support | 

Reviewer: You have a combination of "without support" and "without official support"

They should all either say "without support" or "without commercial support" - I personally prefer the latter.

Author: Going for "without commercial support".

Reviewer: This change wasn't made?

Author: I should really stop drinking while drafting these docs. Fixed now.

[silverstripe/recipe-solr-search](https://github.com/silverstripe/recipe-solr-search) | CMS5 compatible without support | 
[silverstripe/security-extensions](https://packagist.org/packages/silverstripe/security-extensions) | Folded back in core |
[silverstripe/spellcheck](https://packagist.org/packages/silverstripe/spellcheck) | Dropped | 
[silverstripe/sqlite3](https://packagist.org/packages/silverstripe/sqlite3) | CMS5 compatible without support | 
[silverstripe/widgets](https://packagist.org/packages/silverstripe/widgets) | CMS5 compatible without official support | 
[tijsverkoyen/akismet](https://github.com/tijsverkoyen/akismet) | Dropped | 
[tractorcow/classproxy](https://github.com/tractorcow/classproxy) | CMS5 compatible without official support | 
[tractorcow/silverstripe-proxy-db](https://packagist.org/packages/tractorcow/silverstripe-proxy-db) | CMS5 compatible without official support |
[undefinedoffset/sortablegridfield](https://packagist.org/packages/undefinedoffset/sortablegridfield) | Dropped | [symbiote/silverstripe-gridfieldextensions](https://packagist.org/packages/symbiote/silverstripe-gridfieldextensions) provides equivalent functionality.

> ## Fixed dependencies {#fixed-depdencies}

Reviewer: ```suggestion
## Fixed dependencies
```
It will default to `#fixed-dependencies` anyway (which is the same, but without the typo)

Author: Will do

> Silverstripe CMS relies on many third party dependencies to implement various features. Silverstripe CMS defines many *fixed dependencies* that provide certainty to Silverstripe CMS projects and minimise the risk of composer conflicts.

Reviewer: ```suggestion
Silverstripe CMS relies on third party dependencies to implement many features. We have defined several [fixed dependencies](/project_governance/major_release_policy/#dependency-management) to provide certainty to Silverstripe CMS project owners and minimise the risk of composer conflicts. These dependencies will not be upgraded to new major versions within the Silverstripe CMS 5 release line.
```

Author: I'll keep everything except the new link. In this context, it's not clear that you'll will be linking to the Major release policy.

An equivalent link will be in the fixed dependency page any way.

Reviewer: It's clear we'll be linking to something that describes what a "fixed dependency" is, which is what that link goes to. But I'm not too fussed.

> We defined a list of *fixed dependencies* for the lifetime of Silverstripe CMS 5. Those dependencies will not be upgraded to new major version within the Silverstripe CMS 5 release line.

Reviewer: ```suggestion
```

> Review Silverstripe CMS 5 fixed dependencies for more details.

Reviewer: ```suggestion
```
With my above suggestion, this paragraph is no longer needed.

  - [Other features](#other-features)
- [Bugfixes](#bugfixes)

## Change to supported modules {#supported-modules}

> Some Silverstripe CMS modules are commercially supported. Silverstripe commits to looking after those modules for the duration of the Silverstripe CMS 5 lifecycle.

Reviewer 2: ```suggestion
Some Silverstripe CMS modules are commercially supported. Silverstripe Ltd. commits to looking after those modules for the duration of the Silverstripe CMS 5 lifecycle.
```

Is it a commitment by Silverstripe Ltd or Silverstripe as the open source org? How would we make clear which one it is if it was the latter?

Reviewer: I would say it's the company (_commercially_ supported meaning we're talking about the commercial entity that is Silverstripe).

We've been trying to consistently use "Silverstripe" to mean the commercial entity and "Silverstripe CMS" to mean the open source product - though I don't know if we've yet needed to have a way to say "Silverstripe the combined forces of the CMS Squad and core committers". I'm not sure how we would make that distinction.

Author: Our internal style guide calls to avoid  "Silverstripe ltd". Honestly, I think calling the company "Silverstripe" was a mistake, but I'm about 15 years too late to this particular argument.

Reviewer 2: But we can still revive the Silverstripe Foundation idea :)

Review the [list Commercially Supported Modules for Silverstripe CMS 5](/Project_Governance/Supported_Modules).

### Module not supported in Silverstripe CMS 5 {#unsupported-modules}
Some modules that were supported in Silverstripe CMS 4 will not be supported in Silverstripe CMS 5. 

> Just because a module is not supported officially, doesn't mean that you shouldn't be using it. Unsupported modules are maintained on a "best-effort" basis. You should take this into consideration when choosing to install an unsupported module in your project.

Reviewer 2: Not sure how to better phrase this but 'unsupported module' sounds worse than 'module not actively supported' or the like. To me, it somehow implies bigger incompatibility rather than just a lack of official support or a guarantee security issues or bugs will be fixed in a reasonable timeframe. Don't want to stretch us too thin on wording like this, we could be going over it forever, but I can easily imagine some white shirt somewhere reading the changelog and thinking ah, we shouldn't use unsupported modules, heard that from security... 
Maybe we need to always use the long version "modules not supported by Silverstripe within the silverstripe namespace" (which is also not always true, e.g. cwp namespace). I think you get the gist of what I'm trying to convey here.

Author: Maybe "community supported"? I think it contrasts nicely with "commercially supported".

Author: I'll updated that header and the content below it to say "community supported"

Reviewer: This doesn't seem to have been done?

Author: Done now.

- The following classes have been moved from `silverstripe/cms` to `silverstripe/admin` and renamed:
  - `SilverStripe\CMS\Controllers\SilverStripeNavigator` is now in `silverstripe/admin` as `SilverStripe\Admin\Navigator\SilverStripeNavigator`
  - `SilverStripe\CMS\Controllers\SilverStripeNavigatorItem` is now in `silverstripe/admin` as `SilverStripe\Admin\Navigator\SilverStripeNavigatorItem`
  - `SilverStripe\CMS\Controllers\SilverStripeNavigatorItem_Unversioned` is now in `silverstripe/admin` as `SilverStripe\Admin\Navigator\SilverStripeNavigatorItem_Unversioned`
- The following classes have been moved from `silverstripe/cms` to `silverstripe/versioned-admin` and renamed:
  - `SilverStripe\CMS\Controllers\SilverStripeNavigatorItem_ArchiveLink` is now in `silverstripe/versioned` as `SilverStripe\VersionedAdmin\Navigator\SilverStripeNavigatorItem_ArchiveLink`
  - `SilverStripe\CMS\Controllers\SilverStripeNavigatorItem_LiveLink` is now in `silverstripe/versioned` as `SilverStripe\VersionedAdmin\Navigator\SilverStripeNavigatorItem_LiveLink`

>   - `SilverStripe\CMS\Controllers\SilverStripeNavigatorItem_StageLink` is now in `silverstripe/versioned` as `SilverStripe\VersionedAdmin\Navigator\SilverStripeNavigatorItem_StageLink`

Author: Didn't addapi links here because otherwise you wouldn't be able to see the FQN of the classes.

Reviewer: You do realise you can just put the FQCN in the link right? Like ``[`SilverStripe\VersionedAdmin\Navigator\SilverStripeNavigatorItem_ArchiveLink`](api:SilverStripe\VersionedAdmin\Navigator\SilverStripeNavigatorItem_ArchiveLink)``

So if that's the only reason, we should add the link. If you didn't want to do it because it gets messy (which is what I understood from our chat yesterday), then that's totally fine.

Author: It does get messy. I'd rather keep it the way it is.

- `MailTransport`, which used the php native `mail()` function, is no longer present in CMS 5. This is because [Symfony considers `mail()` to be insecure](https://github.com/swiftmailer/swiftmailer/issues/866#issuecomment-289291228).
- If your site has a custom email configuration e.g. SMTP configuration, this will need to be updated, as the configuration has changed from Silverstripe yml to a much more flexible and standardised [DSN string](https://symfony.com/doc/current/mailer.html#transport-setup). See [the email documentation](/developer_guides/email/) for more details. The configuration for email has changed completely - read the updated documentation carefully.
- The following return types where changed on the following methods in [`Email`](api:SilverStripe\Control\Email\Email).
  - Used to return a `string`, now returns an nullable [`Address`](https://symfony.com/doc/current/mailer.html#email-addresses) object.
    - `getReturnPath()`
    - `getSender()`
  - Used to return an array of strings, now returns an array of `Address` objects.
    - `getFrom()`
    - `getTo()`
    - `getBCC()` - note - changed casing to `getBcc()`

>     - `getCC()` - note - changed casing to `getCc()`

Author: Those methods are inherited from Symfony and don't show up in our API doc. So didn't add `api` links here.

TinyMCE has been upgraded up two major versions from 4 to 6. The API for [`HTMLEditorConfig`](api:SilverStripe\Forms\HTMLEditor\HTMLEditorConfig) and [`TinyMCEConfig`](api:SilverStripe\Forms\HTMLEditor\TinyMCEConfig) haven't changed at all, but there have been some changes to plugins and options which may affect your projects. Here are some notable changes:

- The `styleselect` option for selecting a css style has changed to `styles`
- The `spellchecker` plugin is now a premium "[Spell Checker Pro](https://www.tiny.cloud/docs/tinymce/6/introduction-to-tiny-spellchecker/)" plugin through TinyMCE cloud (though the `browser_spellcheck` options still works and is enabled by default)
- The "paste from Microsoft Word" functionality is now part of a premium "[PowerPaste](https://www.tiny.cloud/docs/tinymce/6/introduction-to-powerpaste/)" plugin through TinyMCE cloud

>   - There is an [open call for maintainers](https://github.com/tinymce/tinymce/discussions/7487) for converting the old functionality into a new open source plugin

Author: This seems to be a somewhat pointless thing to point out here.

Reviewer 3: These are changes that, if not pointed out, are likely to trip people up when upgrading. Having custom tinymce config is pretty common, so pointing out changes in that config which are likely to trip people up seems like exactly the sort of thing we should include

Reviewer: I disagree - I think it's worth mentioning, as it is in my prior experience a commonly requested feature. It's worth pointing out that while the old plugin isn't available there is an option to update and maintain it themselves.
Clear upgrade path sometimes means just giving people options, and in this case there's this option which is an alternative to using the pro plugin (which we haven't even tried ourselves and don't know if it's fully compatible with the way we're loading the wysiwyg).

Author: restored

## Front-end build stack upgrades {#front-end}

> We've upgraded the front-end build stack for the CMS, along with most of the javascript dependencies. You only need to worry about this section if you have implemented customs "in-CMS" features with Javascript or CSS.

Reviewer: This should go _below_ the info block, where it was before. The info block _must_ come first, as it gives context as to what "front-end" means in this entire section of the changelog.

> We've upgraded the front-end build stack for the CMS, along with most of the javascript dependencies. You only need to worry about this section if you have implemented customs "in-CMS" features with Javascript or CSS.

Reviewer: ```suggestion
We've upgraded the front-end build stack for the CMS, along with most of the javascript dependencies.
```
"in-CMS" sounds weird to me.... I don't think that sentence is necessary though. If you do want to keep it, put it inside the info block which is here specifically to clarify what "front-end" means and therefore what this section applies to (i.e. only stuff in the CMS)

[warning]
Because this can be controlled with configuration, it is best practice to avoid explicitly expecting a trailing slash to either be present or be omitted.

In PHP you can use methods like [`Controller::join_links()`](api:SilverStripe\Control\Controller::join_links()) or [`Controller::normaliseTrailingSlash()`](api:SilverStripe\Control\Controller::normaliseTrailingSlash()). In JavaScript in your CMS customisations, we recommend using using the `joinUrlPaths()` utility function `silverstripe/admin` which you can access with `import { joinUrlPaths } from 'lib/urls;` if your project uses [@silverstripe/webpack-config](https://www.npmjs.com/package/@silverstripe/webpack-config). In your templates, you should use appropriate methods from your controller or model such as [`SiteTree::Link()`](api:SilverStripe\CMS\Model\SiteTree::Link()) which uses `Controller::join_links()` under the hood.
[/warning]

> The [CanonicalURLMiddleware](api:SilverStripe\Control\Middleware\CanonicalURLMiddleware) will also, by default, redirect traffic to include or omit the trailing slash according to the above configuration. By default, this means that traffic directed to `/about-us/` will be redirected to `/about-us`. You can disable this behaviour with the following YML configuration:

Reviewer: ```suggestion
The [`CanonicalURLMiddleware`](api:SilverStripe\Control\Middleware\CanonicalURLMiddleware) will also, by default, redirect traffic to include or omit the trailing slash according to the above configuration. By default, this means that traffic directed to `/about-us/` will be redirected to `/about-us`. You can disable this behaviour with the following YML configuration:
```

When invoking an extension hook (e.g. via `extend()`), methods prefixed with "extend" will take precedence. i.e. if an `Extension` class has a `onAfterWrite()` method and an `extendOnAfterWrite()` method and I call `$this->extend('onAfterWrite')` - the `extendOnAfterWrite()` method on that `Extension` will be called, and `onAfterWrite()` will not.  
This empowers advanced `Extension` functionality such as [Versioned::canPublish()](api:SilverStripe\Versioned\Versioned::canPublish()) which invokes `$owner->extendedCan('canPublish')` but doesn't result in a cycle, because the same class also implements [extendCanPublish()](api:SilverStripe\Versioned\Versioned::extendCanPublish()).

### Other new features {#other-features}

> - [DataObject::get_one()][api:SilverStripe\ORM\DataObject::get_one()] can now be called directly from subclasses of `DataObject` without passing in a class as the first argument (e.g. `SiteTree::get_one(filter: ['Title:startsWith' => 'About'])`).

Reviewer: ```suggestion
- [`DataObject::get_one()`](api:SilverStripe\ORM\DataObject::get_one()) can now be called directly from subclasses of `DataObject` without passing in a class as the first argument (e.g. `SiteTree::get_one(filter: ['Title:startsWith' => 'About'])`).
```

> - The elemental [TopPage DataExtension](api:DNADesign\Elemental\TopPage\DataExtension) is applied to [BaseElement](api:DNADesign\Elemental\Models\BaseElement) and [`ElementalArea`](api:DNADesign\Elemental\Models\ElementalArea) by default, and the corresponding [`SiteTreeExtension`](api:DNADesign\Elemental\TopPage\SiteTreeExtension) is applied to [`SiteTree`](api:SilverStripe\CMS\Model\SiteTree) by default. This allows you to use `$TopPage` in your elemental block templates as a more performant alternative to `$Page` when referring to the page the element belongs to.

Reviewer: ```suggestion
- The elemental [TopPage DataExtension](api:DNADesign\Elemental\TopPage\DataExtension) is applied to [`BaseElement`](api:DNADesign\Elemental\Models\BaseElement) and [`ElementalArea`](api:DNADesign\Elemental\Models\ElementalArea) by default, and the corresponding [`SiteTreeExtension`](api:DNADesign\Elemental\TopPage\SiteTreeExtension) is applied to [`SiteTree`](api:SilverStripe\CMS\Model\SiteTree) by default. This allows you to use `$TopPage` in your elemental block templates as a more performant alternative to `$Page` when referring to the page the element belongs to.
```
